Checkout: Restart a/b test of a minimal version of the masterbar in checkout

drewblaisdell opened the pull request:
https://github.com/Automattic/wp-calypso/pull/2558 had to be reverted because it added `<CartData />` to the logged-in layout for all users, which caused every client to poll the shopping cart.

This PR starts the test again, and ensures that `<CartData />` is only mounted in `<Layout />` when the user is in checkout, to prevent excessive requests to the shopping cart endpoint.

![screen shot 2016-01-18 at 3 41 51 pm](https://cloud.githubusercontent.com/assets/1130674/12460260/c5c557d2-bf67-11e5-8c9a-80cd419d58d8.png)

**Testing**
- Sign up for a new user through http://calypso.localhost:3000/start
- Run `localStorage.setItem( 'ABTests', '{"checkoutMasterbar_20160125":"original"}' );` in your console.
- Visit http://calypso.localhost:3000/plans and add a plan to your cart.
- Assert that you are redirected to checkout with the full masterbar.
- Run `localStorage.setItem( 'ABTests', '{"checkoutMasterbar_20160125":"minimal"}' );` in your console.
- Refresh the page.
- Assert that you are in checkout with the 'minimal' masterbar.
- Navigate to a page where `<CartData />` isn't mounted, like http://calypso.localhost:3000/pages/:site
- Open the 'Network' pane of the developer tools.
- Assert that your browser is not making repeated requests to the shopping-cart endpoint ([screenshot](https://cloudup.com/cVS5VKl27xo)). You can filter out these requests by typing 'shopping-cart' into the filter input.

- [x] Product review
- [x] Code review

fabianapsimoes commented:
Is it intentional to have the minimal sidebar in the Thank You screen as well?

Looks good apart from that. Product :+1:

stephanethomas commented on client/layout/masterbar/logged-in.jsx:
```diff
@@ -6,6 +6,7 @@ import React from 'react';
 /**
  * Internal dependencies
  */
+import { abtest } from 'lib/abtest';
```
We probably don't need this import.

stephanethomas commented on client/layout/index.jsx:
```diff
@@ -75,6 +78,23 @@ Layout = React.createClass( {
+		if ( 'checkout' === this.props.section && abtest( 'checkoutMasterbar' ) === 'minimal' ) {
+			return (
+				<CartData>
+					<MasterbarMinimal selectedSite={ this.props.sites.getSelectedSite() } />
```
Since this is highly tied to the checkout and the shopping cart, maybe we should name this component `<MasterbarCheckout />` or `<MasterbarLoggedInCheckout />` to make that more explicit. What do you think?

drewblaisdell commented:
@stephanethomas @fabianapsimoes Thanks for reviewing this. I updated the PR to address your notes.

> Is it intentional to have the minimal sidebar in the Thank You screen as well?

Good catch - the full masterbar should be displayed on the 'Thank You' page. I added a commit to address this.

scruffian commented on client/layout/masterbar/checkout.jsx:
```diff
@@ -0,0 +1,32 @@
+/**
```
Do we need to duplicate `client/layout/masterbar/logged-out.jsx`? For the sake of this test I guess it's ok, but I don't think we need to have a different component for this.

drewblaisdell replied:
It depends where we want the logic for determining the cart based on the `cart`/`selectedSite` to live. If we want it in the minimal masterbar component (whatever we name it), I think it should be separate, because otherwise, doing presence checks for `cart`/`selectedSite` seems like a messy way to determine that we're in checkout. Changing the API of this component to accept a `url` prop, moving the logic to determine that URL into (either) layout, seems to solve both issues.

I'll rename this back to `MasterbarMinimal` and update the PR with that change.

gziolo replied:
Nice idea with `MasterbarMinimal` component and composition :+1:

fabianapsimoes commented:
> Good catch - the full masterbar should be displayed on the 'Thank You' page. I added a commit for this.

Hm, I'm still seeing the minimal bar in the Thank You page:

![screenshot-calypso localhost 3000 2016-01-22 11-06-17](https://cloud.githubusercontent.com/assets/1248436/12507786/4778a23e-c0f8-11e5-80b5-486d729f59b3.png)

fabianapsimoes commented:
> Good catch - the full masterbar should be displayed on the 'Thank You' page. I added a commit for this.

Looks good, Drew :+1:

drewblaisdell commented:
I added a commit to address the issue @scruffian raised about duplicating code from the `MasterbarLoggedOut` render() by renaming that component `MasterbarMinimal`, adding a `url` prop, and mounting it in `MasterbarCheckout`.

gziolo commented on client/lib/checkout/index.js:
```diff
@@ -0,0 +1,29 @@
+
+	url = url + siteSlug;
+
+	if ( cartItems.hasRenewalItem( cart ) ) {
```
Shouldn't we put this condition on the top? We would have just to return url from inside to avoid all previous checks that get override by this logic anyway.

drewblaisdell replied:
Good idea. I updated it.

stephanethomas commented:
Code looks good.
